### ISSUE TITLE
Unnecessary resource replace when upgrading from older versions

### ISSUE THREAD

**Issue by ncsibra** [pulumi/pulumi#2829]
Hi,

This is a follow up for the discussion started here: https://github.com/pulumi/pulumi/issues/2781#issuecomment-501652928
@lukehoban I think I have an exact repro for you, my college followed these steps and was able to reproduce the issue.

1. Use these cli/lib/plugin versions, it's important to create the stack with an older version first:
```
➜  pulumi-sandbox (refresh) ✗ pulumi version
v0.17.10
➜  pulumi-sandbox (refresh) ✗ pulumi plugin ls
NAME        KIND      VERSION  SIZE    INSTALLED  LAST USED
aws         resource  0.17.1   157 MB  n/a        2 months ago
```
2. Clone the repo: https://gitlab.com/proemergotech-public/pulumi-sandbox/tree/refresh
3. `yarn install` from the root
4. cd to dliver-sandbox and create a stack, set aws region and `principalUser` in the config, e.g.: `arn:aws:iam::<account_id>:root`
5. `pulumi up`
6. Update aws lib to 0.18.0 in the package.json
7. `yarn install && yarn upgrade` from root
8. Update the pulumi CLI to the latest
9. Export the current stack state
10. `pulumi refresh` from dliver-sandbox, **DO NOT run pulumi up**, it's the key
11. `pulumi up`, 5 resources to update

Now import the previous state, created before the refresh, run `pulumi up`, will be no change, but press yes, then `pulumi refresh`, then `pulumi up` again, nothing will change.
So a refresh without an up messes up something when updating from an older version.

**Comment by lukehoban** [comment]
Thanks for the detailed repro @ncsibra!

**Comment by lukehoban** [comment]
I was able to reproduce.  

The diff at the end is:

```
$ pulumi up
Previewing update (dev):

     Type                         Name                             Plan        Info
     pulumi:pulumi:Stack          dliver-sandbox-dev                           
 +-  ├─ aws:iam:Role              dev-eks-admin-role               replace     [diff: ~name]
 +-  ├─ aws:iam:Policy            dev-eks-admin-policy             replace     [diff: ~name]
 +-  ├─ aws:iam:Role              dev-kubernetes-admin-role        replace     [diff: ~name]
     ├─ network:Vpc               dev                                          
 +-  │  ├─ aws:efs:FileSystem     dev-efs                          replace     [diff: -encrypted,kmsKeyId,performanceMode~creationToken]
 +-  │  ├─ aws:ec2:SecurityGroup  dev-securityGroup-efs            replace     [diff: -egress~ingress,name]
 +-  │  ├─ aws:efs:MountTarget    dev-mountTarget-efs-0            replace     [diff: -ipAddress~fileSystemId,securityGroups]
 +-  │  └─ aws:efs:MountTarget    dev-mountTarget-efs-1            replace     [diff: -ipAddress~fileSystemId,securityGroups]
 +-  └─ aws:iam:PolicyAttachment  dev-eks-admin-policy-attachment  replace     [diff: ~name,policyArn,roles]
 
Resources:
    +-8 to replace
    24 unchanged

Do you want to perform this update? details
  pulumi:pulumi:Stack: (same)
    [urn=urn:pulumi:dev::dliver-sandbox::pulumi:pulumi:Stack::dliver-sandbox-dev]
    ++aws:iam/role:Role: (create-replacement)
        [id=dev-eks-admin-role-338cad0]
        [urn=urn:pulumi:dev::dliver-sandbox::aws:iam/role:Role::dev-eks-admin-role]
      ~ name               : "dev-eks-admin-role-338cad0" => "dev-eks-admin-role-bbe1193"
    +-aws:iam/role:Role: (replace)
        [id=dev-eks-admin-role-338cad0]
        [urn=urn:pulumi:dev::dliver-sandbox::aws:iam/role:Role::dev-eks-admin-role]
      ~ name               : "dev-eks-admin-role-338cad0" => "dev-eks-admin-role-bbe1193"
    ++aws:iam/policy:Policy: (create-replacement)
        [id=arn:aws:iam::153052954103:policy/dev-eks-admin-policy-037f49d]
        [urn=urn:pulumi:dev::dliver-sandbox::aws:iam/policy:Policy::dev-eks-admin-policy]
      ~ name  : "dev-eks-admin-policy-037f49d" => "dev-eks-admin-policy-e8b551b"
    +-aws:iam/policy:Policy: (replace)
        [id=arn:aws:iam::153052954103:policy/dev-eks-admin-policy-037f49d]
        [urn=urn:pulumi:dev::dliver-sandbox::aws:iam/policy:Policy::dev-eks-admin-policy]
      ~ name  : "dev-eks-admin-policy-037f49d" => "dev-eks-admin-policy-e8b551b"
    ++aws:iam/role:Role: (create-replacement)
        [id=dev-kubernetes-admin-role-5c3a38c]
        [urn=urn:pulumi:dev::dliver-sandbox::aws:iam/role:Role::dev-kubernetes-admin-role]
      ~ name               : "dev-kubernetes-admin-role-5c3a38c" => "dev-kubernetes-admin-role-bdcc7b5"
    +-aws:iam/role:Role: (replace)
        [id=dev-kubernetes-admin-role-5c3a38c]
        [urn=urn:pulumi:dev::dliver-sandbox::aws:iam/role:Role::dev-kubernetes-admin-role]
      ~ name               : "dev-kubernetes-admin-role-5c3a38c" => "dev-kubernetes-admin-role-bdcc7b5"
        ++aws:efs/fileSystem:FileSystem: (create-replacement)
            [id=fs-e9b2fe90]
            [urn=urn:pulumi:dev::dliver-sandbox::network:Vpc$aws:efs/fileSystem:FileSystem::dev-efs]
          ~ creationToken  : "dev-efs-c97c2b9" => "dev-efs-e02ad5f"
          - encrypted      : false
          - kmsKeyId       : ""
          - performanceMode: "generalPurpose"
        +-aws:efs/fileSystem:FileSystem: (replace)
            [id=fs-e9b2fe90]
            [urn=urn:pulumi:dev::dliver-sandbox::network:Vpc$aws:efs/fileSystem:FileSystem::dev-efs]
          ~ creationToken  : "dev-efs-c97c2b9" => "dev-efs-e02ad5f"
          - encrypted      : false
          - kmsKeyId       : ""
          - performanceMode: "generalPurpose"
    --aws:iam/policyAttachment:PolicyAttachment: (delete-replaced)
        [id=dev-eks-admin-policy-attachment-0e589c7]
        [urn=urn:pulumi:dev::dliver-sandbox::aws:iam/policyAttachment:PolicyAttachment::dev-eks-admin-policy-attachment]
    +-aws:iam/policyAttachment:PolicyAttachment: (replace)
        [id=dev-eks-admin-policy-attachment-0e589c7]
        [urn=urn:pulumi:dev::dliver-sandbox::aws:iam/policyAttachment:PolicyAttachment::dev-eks-admin-policy-attachment]
      ~ name     : "dev-eks-admin-policy-attachment-0e589c7" => "dev-eks-admin-policy-attachment-c167be2"
      ~ policyArn: "arn:aws:iam::153052954103:policy/dev-eks-admin-policy-037f49d" => output<string>
      ~ roles    : [
          ~ [0]: "dev-eks-admin-role-338cad0" => output<string>
        ]
    ++aws:iam/policyAttachment:PolicyAttachment: (create-replacement)
        [id=dev-eks-admin-policy-attachment-0e589c7]
        [urn=urn:pulumi:dev::dliver-sandbox::aws:iam/policyAttachment:PolicyAttachment::dev-eks-admin-policy-attachment]
      ~ name     : "dev-eks-admin-policy-attachment-0e589c7" => "dev-eks-admin-policy-attachment-c167be2"
      ~ policyArn: "arn:aws:iam::153052954103:policy/dev-eks-admin-policy-037f49d" => output<string>
      ~ roles    : [
          ~ [0]: "dev-eks-admin-role-338cad0" => output<string>
        ]
        ++aws:ec2/securityGroup:SecurityGroup: (create-replacement)
            [id=sg-0198404b8984cb517]
            [urn=urn:pulumi:dev::dliver-sandbox::network:Vpc$aws:ec2/securityGroup:SecurityGroup::dev-securityGroup-efs]
          - egress             : []
          ~ ingress            : [
              ~ [0]: {
                      ~ __defaults    : [
                          + [0]: "self"
                        ]
                      - description   : ""
                      - ipv6CidrBlocks: []
                      - prefixListIds : []
                      - securityGroups: []
                    }
            ]
          ~ name               : "dev-securityGroup-efs-4b6da85" => "dev-securityGroup-efs-a16dd43"
        +-aws:ec2/securityGroup:SecurityGroup: (replace)
            [id=sg-0198404b8984cb517]
            [urn=urn:pulumi:dev::dliver-sandbox::network:Vpc$aws:ec2/securityGroup:SecurityGroup::dev-securityGroup-efs]
          - egress             : []
          ~ ingress            : [
              ~ [0]: {
                      ~ __defaults    : [
                          + [0]: "self"
                        ]
                      - description   : ""
                      - ipv6CidrBlocks: []
                      - prefixListIds : []
                      - securityGroups: []
                    }
            ]
          ~ name               : "dev-securityGroup-efs-4b6da85" => "dev-securityGroup-efs-a16dd43"
        --aws:efs/mountTarget:MountTarget: (delete-replaced)
            [id=fsmt-021a9f7b]
            [urn=urn:pulumi:dev::dliver-sandbox::network:Vpc$aws:efs/mountTarget:MountTarget::dev-mountTarget-efs-0]
        +-aws:efs/mountTarget:MountTarget: (replace)
            [id=fsmt-021a9f7b]
            [urn=urn:pulumi:dev::dliver-sandbox::network:Vpc$aws:efs/mountTarget:MountTarget::dev-mountTarget-efs-0]
          ~ fileSystemId  : "fs-e9b2fe90" => output<string>
          - ipAddress     : "10.151.0.56"
          ~ securityGroups: [
              ~ [0]: "sg-0198404b8984cb517" => output<string>
            ]
        --aws:efs/mountTarget:MountTarget: (delete-replaced)
            [id=fsmt-031a9f7a]
            [urn=urn:pulumi:dev::dliver-sandbox::network:Vpc$aws:efs/mountTarget:MountTarget::dev-mountTarget-efs-1]
        ++aws:efs/mountTarget:MountTarget: (create-replacement)
            [id=fsmt-021a9f7b]
            [urn=urn:pulumi:dev::dliver-sandbox::network:Vpc$aws:efs/mountTarget:MountTarget::dev-mountTarget-efs-0]
          ~ fileSystemId  : "fs-e9b2fe90" => output<string>
          - ipAddress     : "10.151.0.56"
          ~ securityGroups: [
              ~ [0]: "sg-0198404b8984cb517" => output<string>
            ]
        +-aws:efs/mountTarget:MountTarget: (replace)
            [id=fsmt-031a9f7a]
            [urn=urn:pulumi:dev::dliver-sandbox::network:Vpc$aws:efs/mountTarget:MountTarget::dev-mountTarget-efs-1]
          ~ fileSystemId  : "fs-e9b2fe90" => output<string>
          - ipAddress     : "10.151.1.233"
          ~ securityGroups: [
              ~ [0]: "sg-0198404b8984cb517" => output<string>
            ]
        ++aws:efs/mountTarget:MountTarget: (create-replacement)
            [id=fsmt-031a9f7a]
            [urn=urn:pulumi:dev::dliver-sandbox::network:Vpc$aws:efs/mountTarget:MountTarget::dev-mountTarget-efs-1]
          ~ fileSystemId  : "fs-e9b2fe90" => output<string>
          - ipAddress     : "10.151.1.233"
          ~ securityGroups: [
              ~ [0]: "sg-0198404b8984cb517" => output<string>
            ]
    --aws:iam/policy:Policy: (delete-replaced)
        [id=arn:aws:iam::153052954103:policy/dev-eks-admin-policy-037f49d]
        [urn=urn:pulumi:dev::dliver-sandbox::aws:iam/policy:Policy::dev-eks-admin-policy]
        --aws:efs/fileSystem:FileSystem: (delete-replaced)
            [id=fs-e9b2fe90]
            [urn=urn:pulumi:dev::dliver-sandbox::network:Vpc$aws:efs/fileSystem:FileSystem::dev-efs]
    --aws:iam/role:Role: (delete-replaced)
        [id=dev-eks-admin-role-338cad0]
        [urn=urn:pulumi:dev::dliver-sandbox::aws:iam/role:Role::dev-eks-admin-role]
    --aws:iam/role:Role: (delete-replaced)
        [id=dev-kubernetes-admin-role-5c3a38c]
        [urn=urn:pulumi:dev::dliver-sandbox::aws:iam/role:Role::dev-kubernetes-admin-role]
        --aws:ec2/securityGroup:SecurityGroup: (delete-replaced)
            [id=sg-0198404b8984cb517]
            [urn=urn:pulumi:dev::dliver-sandbox::network:Vpc$aws:ec2/securityGroup:SecurityGroup::dev-securityGroup-efs]
```

A subset of the diff in the stack exports is:

```diff
--- stack-old.json	2019-06-14 17:51:31.000000000 -0700
+++ stack-new.json	2019-06-14 17:53:15.000000000 -0700
@@ -1,1265 +1,1394 @@
 {
     "version": 3,
     "deployment": {
         "manifest": {
-            "time": "2019-06-14T17:49:01.181244-07:00",
-            "magic": "a29e66401bf7f3d81c245a3ad28de67d8a633a365aea652e93dc0594b95d5326",
-            "version": "v0.17.10"
+            "time": "2019-06-14T17:51:54.017077-07:00",
+            "magic": "ac764f403142e4dab0729449bfae6db7a2db4774bb1807ef138d7c141976d49b",
+            "version": "v0.17.17"
+        },
+        "secrets_providers": {
+            "type": "service",
+            "state": {
+                "url": "https://api.pulumi.com",
+                "owner": "lukehoban",
+                "project": "dliver-sandbox",
+                "stack": "dev"
+            }
         },
         "resources": [
             {
                 "urn": "urn:pulumi:dev::dliver-sandbox::pulumi:pulumi:Stack::dliver-sandbox-dev",
                 "custom": false,
                 "type": "pulumi:pulumi:Stack",
                 "outputs": {
                     "output": {
                         "efsId": "fs-e9b2fe90",
                         "vpc": {
                             "cidrBlock": "10.151.0.0/16",
                             "fixedIpSubnetIds": [
                                 "subnet-07ab9271a4429d40b",
                                 "subnet-00c07aae85bb0634d"
                             ],
                             "id": "vpc-0757c7f259fbd23f2",
                             "privateIpSubnetIds": [
                                 "subnet-01961458ea6dad32c",
                                 "subnet-00d578aa67bc890db"
                             ],
                             "publicIpSubnetIds": [
                                 "subnet-0924e5eb8f5fa882c",
                                 "subnet-060ce313f5198fa01"
                             ]
                         }
                     }
                 }
             },
             {
                 "urn": "urn:pulumi:dev::dliver-sandbox::pulumi:providers:aws::default",
                 "custom": true,
                 "id": "16447ccb-af6e-4b23-a253-9707d106d7fe",
                 "type": "pulumi:providers:aws",
                 "inputs": {
                     "region": "us-east-2"
                 }
             },
             {
                 "urn": "urn:pulumi:dev::dliver-sandbox::network:Vpc::dev",
                 "custom": false,
                 "type": "network:Vpc",
                 "parent": "urn:pulumi:dev::dliver-sandbox::pulumi:pulumi:Stack::dliver-sandbox-dev"
             },
             {
                 "urn": "urn:pulumi:dev::dliver-sandbox::aws:iam/role:Role::dev-eks-admin-role",
                 "custom": true,
                 "id": "dev-eks-admin-role-338cad0",
                 "type": "aws:iam/role:Role",
                 "inputs": {
+                    "__defaults": [],
                     "assumeRolePolicy": "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Effect\":\"Allow\",\"Principal\":{\"AWS\":\"arn:aws:iam::153052954103:root\"},\"Action\":\"sts:AssumeRole\",\"Condition\":{}}]}",
+                    "description": "",
                     "forceDetachPolicies": false,
                     "maxSessionDuration": 3600,
                     "name": "dev-eks-admin-role-338cad0",
-                    "path": "/"
+                    "path": "/",
+                    "tags": {}
                 },
                 "outputs": {
                     "arn": "arn:aws:iam::153052954103:role/dev-eks-admin-role-338cad0",
                     "assumeRolePolicy": "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Effect\":\"Allow\",\"Principal\":{\"AWS\":\"arn:aws:iam::153052954103:root\"},\"Action\":\"sts:AssumeRole\",\"Condition\":{}}]}",
                     "createDate": "2019-06-15T00:44:11Z",
                     "description": "",
                     "forceDetachPolicies": false,
                     "id": "dev-eks-admin-role-338cad0",
                     "maxSessionDuration": 3600,
                     "name": "dev-eks-admin-role-338cad0",
                     "path": "/",
                     "tags": {},
                     "uniqueId": "AROASHIVKXX33EVN74EJD"
                 },
                 "parent": "urn:pulumi:dev::dliver-sandbox::pulumi:pulumi:Stack::dliver-sandbox-dev",
                 "provider": "urn:pulumi:dev::dliver-sandbox::pulumi:providers:aws::default::16447ccb-af6e-4b23-a253-9707d106d7fe",
                 "propertyDependencies": {
                     "assumeRolePolicy": null
                 }
             },
             {
                 "urn": "urn:pulumi:dev::dliver-sandbox::aws:iam/role:Role::dev-kubernetes-admin-role",
                 "custom": true,
                 "id": "dev-kubernetes-admin-role-5c3a38c",
                 "type": "aws:iam/role:Role",
                 "inputs": {
+                    "__defaults": [],
                     "assumeRolePolicy": "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Effect\":\"Allow\",\"Principal\":{\"AWS\":\"arn:aws:iam::153052954103:root\"},\"Action\":\"sts:AssumeRole\",\"Condition\":{}}]}",
+                    "description": "",
                     "forceDetachPolicies": false,
                     "maxSessionDuration": 3600,
                     "name": "dev-kubernetes-admin-role-5c3a38c",
-                    "path": "/"
+                    "path": "/",
+                    "tags": {}
                 },
                 "outputs": {
                     "arn": "arn:aws:iam::153052954103:role/dev-kubernetes-admin-role-5c3a38c",
                     "assumeRolePolicy": "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Effect\":\"Allow\",\"Principal\":{\"AWS\":\"arn:aws:iam::153052954103:root\"},\"Action\":\"sts:AssumeRole\",\"Condition\":{}}]}",
                     "createDate": "2019-06-15T00:44:11Z",
                     "description": "",
                     "forceDetachPolicies": false,
                     "id": "dev-kubernetes-admin-role-5c3a38c",
                     "maxSessionDuration": 3600,
                     "name": "dev-kubernetes-admin-role-5c3a38c",
                     "path": "/",
                     "tags": {},
                     "uniqueId": "AROASHIVKXX3T6RF4OW65"
                 },
                 "parent": "urn:pulumi:dev::dliver-sandbox::pulumi:pulumi:Stack::dliver-sandbox-dev",
                 "provider": "urn:pulumi:dev::dliver-sandbox::pulumi:providers:aws::default::16447ccb-af6e-4b23-a253-9707d106d7fe",
                 "propertyDependencies": {
                     "assumeRolePolicy": null
                 }
             },
```

**Comment by lukehoban** [comment]
So - the problem here is that the ` "__defaults": [],` entries are getting added here, but do not include the correct entries for the defaults - in particular `name` for the case above.  

It is possible to "fix" this manually by removing every `"__defaults": [],` line from the checkpoint, then importing the resulting checkpoint, then re-running `update`.  I confirmed that doing that results in:

```
$ pulumi up
Previewing update (dev):

     Type                 Name                Plan     
     pulumi:pulumi:Stack  dliver-sandbox-dev           
 
Resources:
    32 unchanged
```

as expected.

After that update, the checkpoint now includes the correct values for `__defaults`:

```
"inputs": {
                    "__defaults": [
                        "name",
                        "path"
                    ],
                    "name": "dev-eks-admin-policy-037f49d",
                    "path": "/",
                    "policy": "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Effect\":\"Allow\",\"Action\":[\"iam:GetRole\",\"iam:PassRole\",\"eks:*\"],\"Resource\":\"*\"}]}"
                },
```

This is indeed a bug in how `__defaults` gets populated along this upgrade path - and we'll need to fix that.  Hopefully the above manual steps allow you to workaround this bug until we can fix the underlying issue.

**Comment by lukehoban** [comment]
One additional note - we believe this will only happen when `refresh` is the first operation that is run after upgrading to a provider version which populates `__defaults`.  As you noted, if you run `update` prior to `refresh`, this issue will not reproduce as correct `__defaults` will be added.

**Comment by ncsibra** [comment]
Thanks for checking, this happened in our sandbox environment, we can destroy and rebuild any time, so no problem.
I didn't know what's the root cause, if it may happen on any provider update or just this one etc. and I thought this could be really dangerous if happens during a CI operation or any other automated script.

**Comment by lukehoban** [comment]
> I didn't know what's the root cause, if it may happen on any provider update or just this on

This would only happen across the specific version boundary where `__defaults` were added about 2 months ago (and only in the case that refresh was the first operation done after upgrading across that boundary). The fix in pulumi/pulumi-terraform#391 should fix this even for that case.
